stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +112,331 @@ internal func _tryNormalize(
+}
+
+private func fastFill(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  _ outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int) {
+  // Quick check if a scalar is NFC and a segment starter
+  @inline(__always) func isNFCStarter(_ scalar: Unicode.Scalar) -> Bool {
+    // Fast-path: All scalars up through U+02FF are NFC and have boundaries
+    // before them
+    if scalar.value < 0x300 { return true }
+
+    // Otherwise, consult the properties
+    return scalar._hasNormalizationBoundaryBefore && scalar._isNFCQCYes
+  }
+  
+  let outputBufferThreshold = outputBuffer.count - 4
+  
+  // TODO: Additional fast-path: All CCC-ascending NFC_QC segments are NFC
+  // TODO: Just freakin do normalization and don't bother with ICU
+  var outputCount = 0
+  let outputEnd = outputBufferThreshold
+  var inputCount = 0
+  let inputEnd = sourceBuffer.count
+  while inputCount < inputEnd && outputCount < outputEnd {
+    // TODO: Slightly faster code-unit scan for latiny (<0xCC)
+
+    // Check scalar-based fast-paths
+    let (scalar, len) = _decodeScalar(sourceBuffer, startingAt: inputCount)
+    _internalInvariant(inputCount &+ len <= inputEnd)
+
+    if _slowPath(
+         !sourceBuffer.hasNormalizationBoundary(before: inputCount &+ len)
+      || !isNFCStarter(scalar)
+    ) {
+      break
+    }
+    inputCount &+= len
+
+    for cu in UTF8.encode(scalar)._unsafelyUnwrappedUnchecked {
+      outputBuffer[outputCount] = cu
+      outputCount &+= 1
+    }
+
+    _internalInvariant(inputCount == outputCount,
+      "non-normalizing UTF-8 fast path should be 1-to-1 in code units")
+  }
+  return (inputCount, outputCount)
+}
+
+// Transcodes a single segment from the source buffer to the outputBuffer as UTF16
+private func transcodeSegmentToUTF16(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
+    
+    if scalar._hasNormalizationBoundaryBefore && readIndex != 0 {
+      break
+    }
+    
+    readIndex += length
+    
+    for cu in scalar.utf16 {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex += 1
+      } else {
+        return nil
+      }
+    }
+  }
+  
+  return (readIndex, writeIndex)
+}
+
+//transcodes the UTF16 segment stored in soureceBuffer into the outputBuffer as UTF8
+private func transcodeSegmentToUTF8(
+  _ sourceBuffer: UnsafeBufferPointer<UInt16>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
```

Comment:
We can't make assumptions about encoding validity for non-native Strings. We error-correct prior to calling ICU, so this happens to be safe, but we should clarify this more in code.

Probably, call this `transcodeValidUTF16ToUTF8`, have some `_internalInvariant` checks that any surrogate is properly paired, update comment on what the return value is, perhaps with tuple labels.

Reply:
ping?